Numeric testcase for precision overflow triggers known protocol violation

bsatarupa opened the pull request:
### Description

Before Fix:
Numeric testcase for precision overflow triggers known protocol violation

After Fix:
Postgres can store numeric data upto 1000 digits but SQL server has a maximum limit of 38 digits. Since, adding the maximum numeric precision check on TDS sender side results into protocol violation, have addressed the issue on engine side by calculating exact numeric digits from Postgres numeric-blocks format and reporting error whenever #digits > TDS_MAX_NUMERIC_PRECISION(i.e. 38).

Task: BABEL-3179
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1720,3 +1726,45 @@ preserve_view_constraints_from_base_table(ColumnDef  *col, Oid tableOid, AttrNum
+	 * Ex. - in 12345.12345 var is of type struct NumericVar; first_block = var->digits[0]= 1,
+	 * var->digits[1] = 2345, var->digits[2] = 1234,
+	 * var->digits[3] = 5000; numeric_base = 4, var->ndigits = #numeric blocks i.e., 4,
+	 * var->weight = 2, var->dscale = 5
```
if take this example : total_digit_count = (2 + 1) * 4 = 12
+1 for first block and +1 for dscale leads to final 14 digits, can you explain how is that correct?

bsatarupa replied:
Breakpoint 2, numeric_sum (fcinfo=<optimized out>) at numeric.c:6010
6010            if ((*detect_numeric_overflow_hook)(sumX_var.weight, sumX_var.dscale, sumX_var.digits[0], DEC_DIGITS))
(gdb) p sumX_var
$1 = {ndigits = 4, weight = 1, sign = 0, dscale = 5, buf = 0x7f1e16330178, digits = 0x7f1e1633017c}

weight = 1; I will rectify my comment in code

bsatarupa replied:
similar calculation of fully filled block digits from numeric_base and weight can be found in numeric() in engine; have added extra calculation for partially filled 1st block and some checks whether last block was partially filled here only.

Commit:
Numeric testcase for precision overflow triggers known protocol violation

Fix:
Postgres can store numeric data upto 1000 digits but SQL server has a maximum limit of 38 digits. Since, adding the maximum numeric precision check on TDS sender side results into protocol violation, have addressed the issue on engine side by calculating exact numeric digits from Postgres numeric-blocks format and reporting error whenever #digits > TDS_MAX_NUMERIC_PRECISION(i.e. 38).

Task: BABEL-3179
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>
